src/adapter/src/coord/sequencer/alter_set_cluster.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use mz_expr::CollectionPlan;
+use mz_sql::catalog::ObjectType;
+use mz_sql::plan::AlterSetClusterPlan;
+use std::iter::once;
+
+use crate::catalog::CatalogItem;
+use crate::coord::Coordinator;
+use crate::session::Session;
+use crate::{catalog, AdapterError, ExecuteResponse, TimestampProvider};
+
+impl Coordinator {
+    /// Convert a [`AlterSetClusterPlan`] to a sequence of catalog operators and adjust state.
+    pub(super) async fn sequence_alter_set_cluster(
+        &mut self,
+        session: &Session,
+        AlterSetClusterPlan { id, set_cluster }: AlterSetClusterPlan,
+    ) -> Result<ExecuteResponse, AdapterError> {
+        let entry = self.catalog().get_entry(&id);
+        match entry.item() {
+            CatalogItem::MaterializedView(catalog::MaterializedView {
+                optimized_expr,
+                cluster_id,
+                ..
+            }) => {
+                if !self
+                    .catalog()
+                    .state()
+                    .is_system_schema_specifier(&entry.name().qualifiers.schema_spec)
+                    && !self.is_compute_cluster(set_cluster)
+                {
+                    let cluster_name = self.catalog().get_cluster(set_cluster).name.clone();
+                    return Err(AdapterError::BadItemInStorageCluster { cluster_name });
+                }
+
+                let cluster_id = *cluster_id;
+
+                // Pick the least valid read timestamp as the as-of for the view
+                // dataflow. This makes the materialized view include the maximum possible
+                // amount of historical detail.
+                let id_bundle = self
+                    .index_oracle(set_cluster)
+                    .sufficient_collections(&optimized_expr.depends_on());
+                let as_of = self.least_valid_read(&id_bundle);
```

Comment:
Should we just use the current `upper` of the MV? We don't care about historical detail here, as that would only result in the new MV trying to write data that the existing one has already written. And by using a later `as_of` we can hope that the new MV hydrates faster.

Reply:
I think in this case we shouldn't create a dataflow at all! If the frontier of a dataflow has advanced to the empty frontier, it is dropped by the replicas as they have nothing to do anymore. The controller probably still needs to keep around the `CollectionState` for bookkeeping, but I think it should be enough to create a `CollectionState` with an empty `as_of` frontier (or copy/move the existing one) and simply not tell the replicas about a new dataflow.

Reply:
Fixed: when the write frontier is `[]`, we just update the catalog and be done. I'm not 100% sure if we need to migrate collection information, tho, but it works :grin: